File: src/algs4/chpt4/part1/A2/BreadthFirstPaths.java
package algs4.chpt4.part1.A2;

import algs4.chpt4.part1.A0.Graph;
import edu.princeton.cs.algs4.Queue;
import edu.princeton.cs.algs4.Stack;

/**
 * Created by hao on 8/18/16.
 */
public class BreadthFirstPaths {

    private boolean[] marked;
    private int[] edgeTo;

    private final int s;

    public BreadthFirstPaths(Graph G, int s) {
        marked = new boolean[G.V()];
        edgeTo = new int[G.V()];
        this.s = s;
        bfs(G, s);
    }

    private void bfs(Graph G, int s) {
        Queue<Integer> q = new Queue<>();
        marked[s] = true;
        q.enqueue(s);
        while (!q.isEmpty()) {
            int v = q.dequeue();
            for (int w : G.adj(v)) {
                if (!marked[w]) {
                    edgeTo[w] = v;
                    marked[w] = true;
                    q.enqueue(w);
                }
            }
        }
    }

    public boolean hasPathTo(int v) {
        return marked[v];
    }

    public Iterable<Integer> pathTo(int v) {
        if (!hasPathTo(v)) return null;
        Stack<Integer> path = new Stack<>();
        int x = v;
        while (x != s) {
            path.push(x);
            x = edgeTo[x];
        }

        path.push(s);

        return path;
    }

}

File: algs4/javascript/chpt4/BreadthFirstPaths.js
var BreadthFirstPaths = function(Graph, s) {
  this.marked = [];
  this.edgeTo = [];
  this.s = s;
  this.bfs(G, s);
};

BreadthFirstPaths.prototype = {
  constructor: BreadthFirstPaths,

  bfs(G, s) {
    let queue = [];
    marked[s] = true;
    queue.push(s);
    while (queue.length > 0) {
      let v = queue.shift(); // dequeue
      let adjs = G.adj(v);
      for (let w of adjs) {
        if (!marked[w]) {
          queue.push(w);
          edgeTo[w] = v;
          marked[w] = true;
        }
      }
    }
  }
}

export default BreadthFirstPaths;